### PR TITLE
Cleanup trivial reduction workarounds

### PR DESCRIPTION
**Pull request by zasdfgbnm**

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/ir_nodes.cpp [hunk 1]
```diff
@@ -1606,49 +1606,37 @@ IterDomain* IterDomain::merge(IterDomain* outer, IterDomain* inner) {
...
 
+  if (outer->isBroadcast()) {
+    itype = inner->getIterType();
+    goto itype_infer_finished;
+  }
+
+  TORCH_CHECK(
+      false, "Merging IterDomains requires that their iteration types match.");
+
+itype_infer_finished:
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/ir_nodes.cpp]: Can this be just written with `if () {} else if () {} ...`?

File: torch/csrc/jit/codegen/cuda/ir_nodes.cpp [hunk 2]
```diff
@@ -1606,49 +1606,37 @@ IterDomain* IterDomain::merge(IterDomain* outer, IterDomain* inner) {
...
-          (outer->isTrivialReduction() && !inner->isReduction()),
-      "Merging IterDomains requires that their iteration types match.");
-  TORCH_CHECK(
-      (outer->isGather() && inner->isGather()) ||
-          (!outer->isGather() && !inner->isGather()),
-      "Merging gather and non-gather domains is not supported.");
-
-  TORCH_CHECK(
-      !outer->isStride() && !inner->isStride(),
-      "No support for merging stride domains");
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/ir_nodes.cpp:1621]: Why are these removed? They still matter, right?

File: torch/csrc/jit/codegen/cuda/ir_nodes.cpp [hunk 3]
```diff
@@ -1606,49 +1606,37 @@ IterDomain* IterDomain::merge(IterDomain* outer, IterDomain* inner) {
...
 
   Val* merged_id_size = mul(outer->extent(), inner->extent());
 
   IterType itype = outer->getIterType();
 
-  if (outer->isBroadcast() && inner->isBroadcast()) {
-    itype = IterType::Broadcast;
+  if (inner->getIterType() == itype) {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/ir_nodes.cpp]: Are there any logical changes here? Or does the new version implement the same logic just in a different way? Looks like it's the latter, but if that's the case, it seems to me the previous version is easier to understand. It's more verbose, but it makes the conditions explicit, which I think really helpful to understand the underlying logic.

File: torch/csrc/jit/codegen/cuda/test/test_gpu.cpp [hunk 4]
```diff
@@ -7853,6 +7853,70 @@ TEST_F(NVFuserTest, FusionReductionScheduler_CUDA) {
       lparams);
 }
 
+TEST_F(NVFuserTest, FusionReductionWithTrivial_CUDA) {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/test/test_gpu.cpp:7860]: Can you please add a brief comment on what is checked by this test?

**Review by naoyam — COMMENTED** [review]
I don't exactly remember, but is this avoidance of trivial reductions no longer necessary?

File: torch/csrc/jit/codegen/cuda/ir_nodes.cpp (hunk 3 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/ir_nodes.cpp]: They are not the same logic. The previous logic does not allow me to merge a (non-trivial) reduction with a broadcasting. And what I am trying to implement here is a pattern matching. I am trying to match the following rules in order. If one rule matches, then we don't proceed with later rules.

```
X + X = X
trivial reduction + X = X
X + trivial reduction = X
broadcasting + X = X
X + broadcasting = X
fail
```
I think this rule could contain all the previous checks, and if-goto seems to be the most natural way to implement this pattern matching in my mind.

If I explain this idea in a comment, something like
```C++
// The itertype inference is a pattern matching of the rules below:
//
//   X + X = X
//   trivial reduction + X = X
//   X + trivial reduction = X
//   broadcasting + X = X
//   X + broadcasting = X
//   fail
//
// The rules are proceeded one by one in order. For each rule, we test if the
// given (outer, inner) matches the pattern. If it does, then we stop
// procceeding and get a result. If we have reached the end without finding any
// matched pattern, then it is a mistake and should be reported.
```
Would you think it will be clear enough, or do you think it is still better to go back to the previous way?

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/ir_nodes.cpp]: OK, I didn't notice merging of reductions and broadcasts. Please add the comments to the code.

I don't agree with the formulation with `goto`. Unless there's no other reasonable way, `goto` shouldn't be used. I think we could easily rewrite this logic with `if` and `else if`, or we could have a separate function that just returns `IterType`, which basically we would just need to replace `goto` with `return`.

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/ir_nodes.cpp]: Also, I think it's worth mentioning that `reduction + broadcast => reduction`, while `trivial reduction + broadcast => broadcast`. The former is obvious as we can't ignore the reduction unless it's trivial. The latter could yield a `trivial reduction`, which might be more consistent. I suppose it doesn't really matter if it's a trivial reduction or a broadcast, so the current logic would be fine, but would be helpful to have a comment on it.

Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/ir_nodes.cpp]: fixed

File: torch/csrc/jit/codegen/cuda/test/test_gpu.cpp (hunk 4 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/test/test_gpu.cpp:7860]: added

**Comment by zasdfgbnm** [comment]
@naoyam I have resolved all your review comments.
> I don't exactly remember, but is this avoidance of trivial reductions no longer necessary?
I did a few tests in the past, and seem that the only thing that was preventing trivial reductions from working was the IdGraph, which was fixed in https://github.com/csarofeen/pytorch/pull/1871

File: torch/csrc/jit/codegen/cuda/ir_nodes.cpp [hunk 5]
```diff
@@ -1597,6 +1597,43 @@ std::vector<IterDomain*> IterDomain::clone(
   return cloned_domains;
 }
 
+IterType inferIterType(IterDomain* i1, IterDomain* i2) {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/ir_nodes.cpp:1600]: @csarofeen I don't have any concern, but please take a look at here just in case.

**Review by naoyam — APPROVED** [review]
LGTM. Thanks for the cleanup.

File: torch/csrc/jit/codegen/cuda/inlining.cpp [hunk 6]
```diff
@@ -153,29 +153,25 @@ size_t MaxPosCalculator::getMaxPosAll(
   return max_pos;
 }
 
-void inlineMost(const std::unordered_set<IterDomain*>& uninlinable_ids) {
-  inlineMost(ir_utils::allTvs(FusionGuard::getCurFusion()), uninlinable_ids);
+void inlineMost() {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/inlining.cpp:156]: For now this seems fine to remove, though I'm open minded we want ID based avoidance of inlining certain dimensions, but we probably want a better interface for that.

File: torch/csrc/jit/codegen/cuda/ir_nodes.cpp (hunk 5 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/ir_nodes.cpp:1600]: Only nit I have is changing the name to understand it only handles resolution through serial, broadcast, and trivial reduced domains. In other words it shouldn't be used for like gather/stride IDs. Nice cleanup though.

**Review by csarofeen — APPROVED** [review]
Nice cleanup, LGTM. Did you try running all the C++ benchmarks just to be sure there isn't anything still broken with trivial reduction merging? I'm only worried about functionality here, not perf.

File: torch/csrc/jit/codegen/cuda/ir_nodes.cpp (hunk 5 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/ir_nodes.cpp:1600]: I think for gather and stride, it applies the first rule `X + X = X`. Do we want to do it differently?

**Comment by zasdfgbnm** [comment]
Finished all benchmarks, no functional issue

**Comment by naoyam** [comment]
Looks like there's a CI failure:

```
00:02:26 [ RUN      ] NVFuserTest.FusionReductionWithTrivial_CUDA
00:02:37 unknown file: Failure
00:02:37 C++ exception with description "aten_output_tensor.allclose( fusion_output_tensor.to(aten_output_tensor.dtype()), tolerance_values.second, tolerance_values.first) INTERNAL ASSERT FAILED at "/opt/pytorch/pytorch/torch/csrc/jit/codegen/cuda/test/test_gpu_validator.h":397, please report a bug to PyTorch. 
00:02:37 
00:02:37 Validation error in output 3 on line 7918 in file /opt/pytorch/pytorch/torch/csrc/jit/codegen/cuda/test/test_gpu.cpp.
00:02:37   Detected abs error of: 14.2804
00:02:37     absolute tolerance was set to 1.30594e-05
00:02:37     and relative tolerance set to 1.30594e-07
00:02:37 Exception raised from testValidate at /opt/pytorch/pytorch/torch/csrc/jit/codegen/cuda/test/test_gpu_validator.h:397 (most recent call first):
00:02:37 frame #0: <unknown function> + 0xc933c (0x7f59d285c33c in /opt/pytorch/pytorch/build/lib/libc10.so)
00:02:37 frame #1: std::function<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > ()>::operator()() const + 0x50 (0x7f59da815722 in /opt/pytorch/pytorch/build/lib/libtorch_cpu.so)
00:02:37 frame #2: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x44 (0x7f59d285b140 in /opt/pytorch/pytorch/build/lib/libc10.so)
00:02:37 frame #3: c10::detail::torchCheckFail(char const*, char const*, unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0x7f (0x7f59d2858e4f in /opt/pytorch/pytorch/build/lib/libc10.so)
00:02:37 frame #4: c10::detail::torchInternalAssertFail(char const*, char const*, unsigned int, char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0x5d (0x7f59d285908a in /opt/pytorch/pytorch/build/lib/libc10.so)
00:02:37 frame #5: <unknown function> + 0x99c027 (0x55b6ea0cd027 in bin/test_jit)
```

**Comment by zasdfgbnm** [comment]
🤯 Just checked locally, this passes on Ampere (3090), but fails on Turing (2080ti)

**Comment by zasdfgbnm** [comment]
The failing shape and reduction dims are:
```C++
  std::vector<std::vector<int64_t>> shapes = {
      /*{-1, -1, 1}, {-1, 1, -1},*/ {1, -1, -1}};
```

```C++
    std::vector<std::vector<int64_t>> reduction_dims = {
        // {0},
        // {1},
        // {2},
        // {0, 1},
        // {0, 2},
        // {1, 2},
        {0, 1, 2},
    };
```

**Comment by zasdfgbnm** [comment]
Also checked that Ampere and Turing are lowered to exactly the same CUDA code. Except that the execution result of Ampere is correct, but for Turing it is wrong
